discovery/sync_manager.go

```diff
@@ -0,0 +1,619 @@
+		// peer and force a full sync with them.
+		case <-m.cfg.FullSyncTicker.Ticks():
+			m.forceFullSync()
+
+		// A new activeSyncer has been received, so we'll either attempt
+		// to start it if the queue is empty or queue it so that we
+		// don't attempt to sync with multiple peers at the same time.
+		case s := <-m.newActiveSyncers:
+			if current == nil {
+				if err := startSyncer(s); err != nil {
+					log.Errorf("Unable to start "+
+						"gossipSyncer(%x): %v", err)
+					continue
+				}
+
+				current = s
+				continue
+			}
+
+			log.Debugf("Queueing next active syncer for peer=%x",
+				s.cfg.peerPub)
+
+			m.pendingActiveSyncers = append(m.pendingActiveSyncers, s)
+
+		// A stale activeSyncer has been received, so we'll need to
+		// remove them from our queue. If we are currently waiting for
+		// its state machine to complete, we'll move on to the next
+		// activeSyncer in the queue.
+		case peer := <-m.staleActiveSyncers:
+			if current != nil && peer == current.cfg.peerPub {
+				startNextSyncer()
+				continue
+			}
+
+			for i, activeSyncer := range m.pendingActiveSyncers {
+				if activeSyncer.cfg.peerPub == peer {
+					m.pendingActiveSyncers = append(
+						m.pendingActiveSyncers[:i],
+						m.pendingActiveSyncers[i+1:]...,
+					)
+					break
+				}
+			}
+
+		// Our current active syncer has reached its terminal
+		// chansSynced state, so we'll proceed to starting the next
+		// pending active syncer if there is one.
+		case <-startNext:
+			log.Debugf("Received chansSynced signal from "+
+				"gossipSyncer(%x)", current.cfg.peerPub)
+
+			startNextSyncer()
+
+		// We've timed out waiting for the current active syncer to
+		// reach its terminal chansSynced state.
+		case <-m.cfg.ActiveSyncerTimeoutTicker.Ticks():
+			// If there are no pending active syncers left, then we
+			// can just reset the timeout for the current one.
+			if len(m.pendingActiveSyncers) == 0 {
+				// TODO(wilmer): Restart syncer state machine?
+				log.Debugf("Rescheduling active syncer timeout "+
+					"for peer=%x", current.cfg.peerPub)
+				continue
+			}
+
+			// Otherwise, we'll move on to the next active syncer in
+			// the list and add the current one to the end of our
+			// queue.
+			log.Debugf("Timed out waiting for active syncer with "+
+				"peer=%x to complete", current.cfg.peerPub)
+
+			m.pendingActiveSyncers = append(
+				m.pendingActiveSyncers, current,
+			)
+
+			startNextSyncer()
+
+		case <-m.quit:
+			return
+		}
+	}
+}
+
+// nextPendingActiveSyncer returns the next activeSyncer pending to be start. If
+// there aren't any, then `nil` is returned.
+func (m *syncManager) nextPendingActiveSyncer() *activeSyncer {
+	switch len(m.pendingActiveSyncers) {
+	case 0:
+		return nil
+	case 1:
+		s := m.pendingActiveSyncers[0]
+		m.pendingActiveSyncers = nil
+		return s
+	default:
+		s := m.pendingActiveSyncers[0]
+		m.pendingActiveSyncers = m.pendingActiveSyncers[1:]
```

Comment:
Should also set `pendingActiveSyncers[0] = nil` to release the dangling reference.

Reply:
Fixed.